src/appengine/handlers/base_handler_flask.py

```diff
@@ -0,0 +1,300 @@
+
+    if values['is_logged_in']:
+      values['switch_account_url'] = make_login_url(request.url)
+      values['logout_url'] = make_logout_url(dest_url=request.url)
+
+    template = _JINJA_ENVIRONMENT.get_template(path)
+
+    self._add_security_response_headers()
+    self.response.headers['Content-Type'] = 'text/html'
+    self.response.data = template.render(values)
+    self.response.status_code = status
+    return self.response
+
+  def before_render_json(self, values, status):
+    """A hook for modifying values before render_json."""
+
+  def render_json(self, values, status=200):
+    """Write JSON response."""
+    self._add_security_response_headers()
+    self.response.headers['Content-Type'] = 'application/json'
+    self.before_render_json(values, status)
+    self.response.data = json.dumps(values, cls=JsonEncoder)
+    self.response.status_code = status
+    return self.response
+
+  def handle_exception(self, exception):
+    """Catch exception and format it properly."""
+    try:
+      status = 500
+      values = {
+          'message': str(exception),
+          'email': helpers.get_user_email(),
+          'traceDump': traceback.format_exc(),
+          'status': status,
+          'type': exception.__class__.__name__
+      }
+      if isinstance(exception, helpers.EarlyExitException):
+        status = exception.status
+        values = exception.to_dict()
+
+      # 4XX is not our fault. Therefore, we hide the trace dump and log on
+      # the INFO level.
+      if 400 <= status <= 499:
+        logging.info(json.dumps(values, cls=JsonEncoder))
+        del values['traceDump']
+      else:  # Other error codes should be logged with the EXCEPTION level.
+        logging.exception(exception)
+
+      if helpers.should_render_json(
+          request.headers.get('accept', ''),
+          self.response.headers.get('Content-Type')):
+        return self.render_json(values, status)
+      if status in (403, 401):
+        return self.render_forbidden(str(exception))
+      return self.render('error.html', values, status)
+    except Exception:
+      self.handle_exception_exception()
+
+  def handle_exception_exception(self):
+    """Catch exception in handle_exception and format it properly."""
+    exception = sys.exc_info()[1]
+    values = {'message': str(exception), 'traceDump': traceback.format_exc()}
+    logging.exception(exception)
+    if helpers.should_render_json(
+        request.headers.get('accept', ''),
+        self.response.headers.get('Content-Type')):
+      return self.render_json(values, 500)
+    return self.render('error.html', values, 500)
+
+  def redirect(self, url, **kwargs):
+    """Check vaid url and redirect to it, if valid."""
+    url = str(url)
+    check_redirect_url(url)
+    return flask_redirect(url, **kwargs)
+
+  def dispatch_request(self, *args, **kwargs):
+    """Dispatch a request and postprocess."""
+    self.response = Response()
```

Comment:
let's avoid this as well. This seems inconsistent with how requests are accessed. One difference here with webapp2 is that flask expects this as a return value in the handler.

So, we can just construct this in self.render without pre-constructing it here, or in the rare cases where that is not used, just construct it directly in the handler.